src/system/SystemLayer.cpp

```diff
@@ -282,28 +282,102 @@ void Layer::StartTimer(uint32_t aMilliseconds, chip::Callback::Callback<> * aCal
  *   @param[in]  aComplete     A pointer to the function called when timer expires.
  *   @param[in]  aAppState     A pointer to the application state object used when timer expires.
  *
- *   @return CHIP_SYSTEM_NO_ERROR On success.
- *   @return CHIP_SYSTEM_ERROR_NO_MEMORY If a timer cannot be allocated.
- *   @return Other Value indicating timer failed to start.
+ *   @return Timer object handle On success.
+ *   @return nullptr if a timer failed to start.
  *
  */
-Error Layer::StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, void * aAppState)
+Timer * Layer::StartTimer(uint32_t aMilliseconds, TimerCompleteFunct aComplete, void * aAppState)
 {
     Error lReturn;
     Timer * lTimer;
 
     this->CancelTimer(aComplete, aAppState);
     lReturn = this->NewTimer(lTimer);
-    SuccessOrExit(lReturn);
+    if (lReturn != CHIP_SYSTEM_NO_ERROR)
+    {
+        return nullptr;
+    }
 
     lReturn = lTimer->Start(aMilliseconds, aComplete, aAppState);
     if (lReturn != CHIP_SYSTEM_NO_ERROR)
     {
         lTimer->Release();
+        return nullptr;
     }
 
-exit:
-    return lReturn;
+    return lTimer;
+}
+
+/**
+ * @brief
+ *   This method extends a one-shot timer.
```

Comment:
No, this method creates a new timer that will fire at the extended point, and cancels the old timer, the one that that caller is using as a handle.

So in particular, given this code:

```
Timer * t = SystemLayer.StartTimer(a, func, nullptr);
// time passes
SystemLayer.ExtendTimer(b, t);
// time passes
SystemLayer.ExtendTimer(c, t);
```
the results of the third call are totally unpredictable. Maybe it'll make it so `func` will run after time `a + b + c` after that first `StartTimer` call.  Maybe it'll extend some other totally unrelated timer and `func` will run after time `a+b`.  Maybe it will just fail because `t` is not a live timer at that point (it got canceled by the first ExtendTimer call, recall, and the only way the other two scenarios happen is if some other timer got allocated in that slot).

The fact that the tests did not catch this suggests the tests are not sufficient.

Reply:
In particular, either ExtendTimer needs to keep the handle it's given working, or it needs to return the new handle.

Reply:
The ExtendTimer always first check the validity of the input timer. so the result of the third call is not unpredictable, it either return error CHIP_SYSTEM_ERROR_TIMER_NOT_FOUND or 'a+b+c'.

Reply:
The first `ExtendTimer` call frees `t` in my example and creates a new timer for time `a+b`.  This new timer may end up being `t`, or might be a different timer in the list, depending on where the first free slot is after `t` is freed.

So when the second `ExtendTimer` call happens, if the second call reused `t` then you get `a+b+c`.  If the second call did _not_ reuse `t`, then you might get `CHIP_SYSTEM_ERROR_TIMER_NOT_FOUND` or you might extend some other random timer by `c` if someone started a timer and it used the free timer `t` points to.